### DISCUSSION TITLE
TFT_SPI compatibility with read/write to Serial UART on RP2040

### DISCUSSION THREAD

**Discussion by SWills73** [Bodmer/TFT_eSPI#3336] [Q&A - General] [answered]
Hi, looking for some community help after spending days trying to resolve and going nuts in the process.

I have a Feather RP2040 with an ST7789 display using TFT_eSPI and want to read data from a Serial port and Deserialize with ArduinoJSON.

The TFT_eSPI works with the display (Perfect!), and I've had the read/write serial stream working, - but **not** at the same time. As soon as I send data to the display, the Serial data stream cuts out.

I've tried to ensure I don't have conflicting PINS, but need to help to understand if this is the issue.

TFT_eSPI User_Setup File:
#define ST7789_DRIVER  
#define TFT_MOSI  19
#define TFT_SCLK  18
#define TFT_CS    13  // Chip select control pin
#define TFT_DC    12  // Data Command control pin
#define TFT_RST   11  // Reset pin (could connect to Arduino RESET pin)
#define RP2040_PIO_SPI 

Feather RP2040 Pin Setup (using the Earle Philhower arduino-pico core)
  Serial2.setRX(25);
  Serial2.setTX(24);

  Serial2.begin(57600);  // initialise serial interface

What Works:
1. TFT_eSPI with ST7789 IPS display 
2. Data streaming when not sending anything to the display in the program loop

Failure Mode:
- conflict of TFT_eSPI setup with the Serial streaming port
- I've tried various Serial RX/TX pins - including 28&29 (TX0/RX0) and 24&25 (TX1/RX1)- I can get the stream to work and deserialize but not with the display output

Struggling to understand the conflict of TFT_eSPI which prevents a correct data stream on Serial

**Comment by SWills73** [comment, marked as answer]
Fixed - I was pushing the TFT_eSPI display update too often so it was blocking the Serial stream; I put a 250ms minimum timer on display updates and it allows the Serial UART to do its thing.

**Comment by vitalyxxl** [comment]
I have a similar situation, but on the ESP32 chip.

1 Display - ST7789. 
2 Cyclic reading of data from the Serial1 (or Serial2) port:

#include <Arduino.h>
#include <SPI.h>
#include <TFT_eSPI.h>

// TFT_DC 32
// TFT_RST 33
// TFT_CS -1
// TFT_MOSI 25
// TFT_SCLK 26
TFT_eSPI tft = TFT_eSPI(); // Creating object (tft.)

Library for reading and parsing data from the Serial port:

#include "HLW8032.h"
HLW8032 HL; // creating object-electricity meter (HL.)
......

tft.init(); // ST7789 DISPLAY INIT
HL.begin(Serial1,27); // SerialPort, Pin for reading.

The display output is excellent, without delays and artifacts. 
But receiving data from the UART port (Serial1, GPIO27) is an error. Packets of the wrong length arrive. Instead of the 24 bytes laid down by the algorithm - any other - most often 9-13 bytes. Accordingly, the contents of even this truncated part do not correspond to the transmission protocol of the microcircuit (HLV8032).

If we do not use the initialization command tft.init(); , then data reception from the port occurs correctly, as laid down by the program without any errors or breaks.
Instead of the TFT_esPAY library, I tried to connect other similar ones. Everything works exactly the same with errors (or rather, it doesn't work)....

Obviously, there is a conflict between SPY and UART.
 But I don't know how to fix it. I ask my colleagues for help. In general, is there any solution? Thanks in advance!

**Comment by KlausLi** [comment]
void setup() {
  Serial.begin(115200);
  while(!Serial) delay(20);
  
  Serial.println("Serial Init.");// "Hello world."
  // ohne flush würds hier nicht gehen , 
  //warum weiss ich nicht , ich weiss es 
  //liegt an der libary TFT_espi und bei 
  //aufruf von tft.begin oder tft.init ist der Serial TOT
  //Hat mich einen Tag gekostet°!°!°
  Serial.flush(); 

  //Der Serial Killer
  tft.begin();

  //Serial ist toto nach TFT init 
  Serial.end();
  while(Serial) delay(20);
  Serial.begin(115200);
  while(!Serial) delay(20);
  Serial.println("SERIAL NEUSTART nach tft INIT, weil danch die serial tot ist, auch ist kein flush mehr nötig, evtl. gibt es Update von TFT_eSpi libary ?");// "Hello world."
  //restart serial end
//Ab jetzt funktioniert wieder alles 
  tft.setRotation(1);     // Adjust Rotation of your screen (0-3)
  tft.fillScreen(TFT_BLACK);
  
gruss Klaus 
http://www.youtube.com/@BavarianSuperGuy
